File: python/voxgig_sekreto/plugins/infisical.py
# The infisical plugin: Infisical. Needs HTTPS. A port of
# typescript/plugins/infisical.ts.
import json
import time

from ..addr import checkaddr
from ..sekreto import SekretoError, envkey
from ..providers import Provider, providerplugin
from .httpjson import fetchjson, tonumber, urlpart


class InfisicalProvider(Provider):
    """Infisical.

    `api.token` reads the secret keyed `API_TOKEN` (Infisical's own
    convention is environment-style keys) at a secret path in one
    environment of a project. Auth is a token, or a universal-auth
    (machine identity) login with clientid/clientsecret.
    """

    def __init__(self, options=None):
        self.opts = options or {}

        # A configured token is kept forever; a universal-auth token carries
        # expiresIn and is renewed shortly before it runs out.
        self.livetoken = None
        self.renewat = float('inf')

    def login(self, addr):
        if self.opts.get('token'):
            return self.opts['token']

        if not self.opts.get('clientid') or not self.opts.get('clientsecret'):
            raise SekretoError('sekreto: infisical: no token and no client credentials')

        status, body = fetchjson(
            'POST',
            addr + '/api/v1/auth/universal-auth/login',
            {'content-type': 'application/json'},
            json.dumps(
                {'clientId': self.opts['clientid'], 'clientSecret': self.opts['clientsecret']},
                separators=(',', ':'),
            ),
        )

        got = (body or {}).get('accessToken')
        if 200 != status or not got:
            raise SekretoError('sekreto: infisical login failed: ' + str(status))

        expires = tonumber((body or {}).get('expiresIn'))
        self.renewat = time.time() + max(expires - 60, 1) if 0 < expires else float('inf')

        return str(got)

    def lookup(self, name):
        addr = (self.opts.get('addr') or 'https://app.infisical.com').rstrip('/')
        checkaddr(addr)

        project = self.opts.get('project') or ''
        environment = self.opts.get('environment') or ''
        if '' == project or '' == environment:
            raise SekretoError('sekreto: infisical: no project/environment')

        if self.livetoken is None or time.time() >= self.renewat:
            self.livetoken = self.login(addr)

        url = (addr + '/api/v3/secrets/raw/' + envkey(name)
               + '?workspaceId=' + urlpart(project)
               + '&environment=' + urlpart(environment)
               + '&secretPath=' + urlpart(self.opts.get('path') or '/'))

        status, body = fetchjson('GET', url, {'authorization': 'Bearer ' + self.livetoken})

        if 404 == status:
            return None

        if 200 != status:
            raise SekretoError('sekreto: infisical error: ' + str(status))

        value = ((body or {}).get('secret') or {}).get('secretValue')
        return None if value is None else str(value)

    def describe(self):
        return ('infisical:' + (self.opts.get('project') or '')
                + '/' + (self.opts.get('environment') or ''))


# The plugin: the `infisical` provider kind, as a voxgig/plugin definition.
infisical = providerplugin('infisical', InfisicalProvider)

File: ruby/lib/voxgig_sekreto/plugins/infisical.rb
# frozen_string_literal: true

# The infisical plugin: Infisical. Needs HTTPS. A port of
# typescript/plugins/infisical.ts, which is canonical.

require 'json'

require_relative '../../voxgig_sekreto'
require_relative '../addr'
require_relative 'httpjson'

module VoxgigSekreto
  # Infisical.
  #
  # `api.token` reads the secret keyed `API_TOKEN` (Infisical's own
  # convention is environment-style keys) at a secret path in one
  # environment of a project. Auth is a token, or a universal-auth
  # (machine identity) login with clientid/clientsecret.
  class InfisicalProvider
    def initialize(opts = nil)
      @opts = opts || {}

      # A configured token is kept forever; a universal-auth token carries
      # expiresIn and is renewed shortly before it runs out.
      @livetoken = nil
      @renewat = Float::INFINITY
    end

    def login(addr)
      return @opts['token'] if VoxgigSekreto.given(@opts['token'])

      unless VoxgigSekreto.given(@opts['clientid']) && VoxgigSekreto.given(@opts['clientsecret'])
        raise SekretoError, 'sekreto: infisical: no token and no client credentials'
      end

      status, body = VoxgigSekreto.fetchjson(
        'POST', addr + '/api/v1/auth/universal-auth/login',
        { 'content-type' => 'application/json' },
        JSON.generate('clientId' => @opts['clientid'], 'clientSecret' => @opts['clientsecret'])
      )

      got = body.is_a?(Hash) ? body['accessToken'] : nil
      unless 200 == status && VoxgigSekreto.given(got)
        raise SekretoError, 'sekreto: infisical login failed: ' + status.to_s
      end

      expires = body['expiresIn'].to_f
      @renewat = 0 < expires ? Time.now.to_f + [expires - 60, 1].max : Float::INFINITY

      got.to_s
    end

    def lookup(name)
      addr = (@opts['addr'] || 'https://app.infisical.com').sub(%r{/\z}, '')
      VoxgigSekreto.checkaddr(addr)

      project = @opts['project'] || ''
      environment = @opts['environment'] || ''
      if '' == project || '' == environment
        raise SekretoError, 'sekreto: infisical: no project/environment'
      end

      @livetoken = login(addr) if @livetoken.nil? || Time.now.to_f >= @renewat

      url = addr + '/api/v3/secrets/raw/' + VoxgigSekreto.envkey(name) +
            '?workspaceId=' + VoxgigSekreto.uriescape(project) +
            '&environment=' + VoxgigSekreto.uriescape(environment) +
            '&secretPath=' + VoxgigSekreto.uriescape(@opts['path'] || '/')

      status, body = VoxgigSekreto.fetchjson('GET', url,
                                             { 'authorization' => 'Bearer ' + @livetoken })

      return nil if 404 == status

      raise SekretoError, 'sekreto: infisical error: ' + status.to_s if 200 != status

      value = body.is_a?(Hash) && body['secret'].is_a?(Hash) ? body['secret']['secretValue'] : nil
      value.nil? ? nil : value.to_s
    end

    def describe
      'infisical:' + (@opts['project'] || '') + '/' + (@opts['environment'] || '')
    end
  end

  module Plugins
    # The plugin: the `infisical` provider kind, as a voxgig/plugin
    # definition.
    INFISICAL = VoxgigSekreto.providerplugin('infisical', lambda { |spec|
      InfisicalProvider.new(spec.slice('addr', 'token', 'clientid', 'clientsecret',
                                       'project', 'environment', 'path'))
    })
  end
end
